### PR TITLE
Add a test for UI-Auth-via-SSO

### PR DESCRIPTION
**Pull request by richvdh**
This adds a relatively complete test of a UI Auth operation using an OIDC IdP (it just stubs out the calls from synapse to the IdP). Most of this stuff wasn't really tested before.

Sorry the diff isn't particularly readable. The idea is to leverage some of the exiting code in `login_via_oidc` to support UIA as well as login: as such I've split out a more generic `auth_via_oidc` which can be used for either UIA or login.

### CODE REVIEW

File: tests/rest/client/v1/utils.py [hunk 1]
```diff
@@ -413,34 +473,97 @@ async def mock_req(method: str, uri: str, data=None, headers=None):
...
+                if tag == "a":
+                    href = attr_dict["href"]
+                    if href:
+                        self.links.append(href)
+
+            def error(_, message):
+                raise AssertionError(message)
+
+        p = ConfirmationPageParser()
+        p.feed(channel.result["body"].decode("utf-8"))
```
Comment (clokep) [inline tests/rest/client/v1/utils.py:562]: I think you can use the next `text_body()` method now? Or do we not guarantee this is a `FakeChannel`?

File: tests/rest/client/v1/utils.py [hunk 2]
```diff
@@ -362,34 +363,93 @@ def login_via_oidc(self, remote_user_id: str) -> JsonDict:
...
-        # param that synapse passes to the IdP via query params, and the cookie that
-        # synapse passes to the client.
-        assert channel.code == 302
-        oauth_uri = channel.headers.getRawHeaders("Location")[0]
-        params = urllib.parse.parse_qs(urllib.parse.urlparse(oauth_uri).query)
-        redirect_uri = "%s?%s" % (
+        # param that synapse passes to the IdP via query params, as well as the cookie
+        # that synapse passes to the client.
+
+        oauth_uri_path, oauth_uri_qs = oauth_uri.split("?", 1)
```
Comment (clokep) [inline tests/rest/client/v1/utils.py:441]: My brain says this should use `urlparse` (and `urlunparse` below), but I don't think it is super important.

File: tests/rest/client/v1/utils.py [hunk 3]
```diff
@@ -362,34 +363,93 @@ def login_via_oidc(self, remote_user_id: str) -> JsonDict:
...
+        Args:
+            remote_user_id: the remote id that the OIDC provider should present
+            client_redirect_url: for a login flow, the client redirect URL to pass to
+                the login redirect endpoint
+            ui_auth_session_id: if set, we will perform a UI Auth flow. The session id
+                of the UI auth.
+
+        Returns:
+            the result of calling the OIDC callback endpoint - which may be a 200, 302
+            or 400 depending on how things went.
```
Comment (clokep) [inline tests/rest/client/v1/utils.py:422]: This seems to return a `FakeChannel`, not just the status code.

**Review by clokep — APPROVED** [review]
LGTM! Thanks for adding some tests!

File: tests/rest/client/v1/utils.py (hunk 1 above)
Comment (richvdh) [reply tests/rest/client/v1/utils.py:562]: yup, true.

File: tests/rest/client/v1/utils.py (hunk 2 above)
Comment (richvdh) [reply tests/rest/client/v1/utils.py:441]: I flipflopped on that a bit, but basically I don't really want to have to reassemble bits of the urlparse output to compare against `oauth_uri_path` on the next line.

**Pull request merged**